Require that the strings are non-empty. (#3487)
modelica/ModelicaSpecification: chapters/annotations.tex

The conversion should generate correct Modelica models using the new version of the library
corresponding to the old version.

> The string arguments for these functions should not be the empty string (except for the string in an empty array of strings).

Reviewer: Two things:
- The decision was to forbid, so _shall_.
- I don't understand how an empty array of strings may contain an empty string.  if the comment has to do with `fill("", 0)`, I don't see the point of trying to clarify.

Author: I have removed this entire part as it was duplicated, and used the previously suggested text.

Author: @henrikt-ma ?

\begin{nonnormative}
Whenever possible tools should preserve the original style of the model, e.g., use of imports.
Conversions should be applied in all places where named element are used in code, including Modelica URIs (for example, in \lstinline!Documentation! annotations).
\end{nonnormative}

> These functions can be called with literal strings or array of strings and vectorize according to \cref{scalar-functions-applied-to-array-arguments}.

Reviewer: Maybe fix grammar as well?
```suggestion
These functions can be called with literal strings or arrays of strings and vectorize according to \cref{scalar-functions-applied-to-array-arguments}.
```

Reviewer: The current formulation leaves the status of something like `{"a" + "b"}` unclear.  Is the real intention to support:
- Literal strings.
- Arrays of literal strings (for example, `{"a", "b"}`).
- Calls to `fill` with a literal string as element value (for example, `fill("", 0)`).

Given this, the meaning of the following would be clear:
- The empty literal string is only allowed when constructing an empty array using `fill`.

Conversions should be applied in all places where named element are used in code, including Modelica URIs (for example, in \lstinline!Documentation! annotations).
\end{nonnormative}

> These functions can be called with literal strings or arrays of strings and vectorize according to \cref{scalar-functions-applied-to-array-arguments}.

Reviewer: I think there must be some missing rules for the arrays of strings.  I get the impression that both `{"a"}` and `fill("", 0)` are examples of allowed arrays, but what about the following?
- `{"a" + "b"}
- `{String(true)}`

Could it be that the intention is to say that the only allowed kind of string expression is string literals, and that arrays of strings may be constructed using `fill` and `array` (possibly as `{…}`) without iterators.

Are there good use cases for allowing `fill` to create non-empty arrays?

Specifying conversions using the old name of a class allows the conversion to be done without access to the old
version of the library (by suitable modifications of the lookup).  Another alternative is to use the old version
of the library during the conversion.

> An empty string would either mean an empty diagnostics, a missing class or component, or a modifier which does not make sense.

Reviewer: It's hard to include this kind of statement about something illegal in a meaningful way.  If it should be included, I think it needs to be more clear why:
```suggestion
Not allowing the empty string just means that it is not possible to express meaningless things such as empty diagnostics, a missing class or component, or a non-sensical modifier.
```

Reviewer 2: Although this sentence is technically correct, I do not see how it helps to clarify anything. In #3459 I was just thinking of convertElement. But it is helpful to make a full list of what cases are accepted and make it normative. 
- Any `"OldClass"`, `"oldElement"`, or `"oldName"` *must* be non-empty.  Should compilers just ignore them or result in an error?
- Should it be allowed to have `"NewClass"` as non-empty? Would it be equivalent to `convertMessage` with and empty failed message or would it be just ignored?
- Arrays for modifiers in `convertModifiers`: either one can be empty. Already explained.
- Empty `"NewName"` is the question of #3459: is equivalent to `convertModifier("OldClass","OldName",fill{"",0})`. It will result in an error if a model that is converted uses `OldClass.OldName` anywhere other than in a modification.

Author: Hm...
One option is to just skip this explanation for now, if the explanation isn't helpful.
Another is to spell out how to avoid using an empty string:

The following are alternatives to using an empty string:
* Instead of empty OldClass, oldElement, oldName (as a form of wild-card) list all of them.
* Instead of empty NewClass or NewElement use convertMessage to indicate why it cannot be converted.
* Instead of empty NewName for convertElement use convertModifiers("OldClass",{"OldName"},fill("",0)) - even if limited to modifications.
* Instead of empty message for convertMessage give a message explaining why.

However, looking at this list it looks more like an annotated reference than something that is needed in the specification, so I think it is easier to just drop this line.

Author: I believe I will just erase it.
Is that ok, @eshmoylova @henrikt-ma (and others)?

Reviewer 2: I am ok with erasing it. But I want a sentence added to convertElement that states that case `convertElement("OldClass","OldName",fill("",0))` is equivalent to `convertModifiers("OldClass",{"OldName"},fill("",0))` and no conversion is applied to any equations.

Author: Ok, I have gladly removed the description of why, and reluctantly added the above.

Specifying conversions using the old name of a class allows the conversion to be done without access to the old
version of the library (by suitable modifications of the lookup).  Another alternative is to use the old version
of the library during the conversion.

> An empty string would mean either mean an empty diagnostics or a missing class, component, or modifier which does not make sense.

Reviewer: Too mean:
```suggestion
An empty string would either mean an empty diagnostics, a missing class or component, or a modifier which does not make sense.
```
